### ISSUE TITLE
Magento 2.3.0-beta18: values in sequence_table column in sales_sequence_meta table get cut-off an cause unique constraint violation on setup:upgrade

### ISSUE THREAD

**Issue by jhruehl** [magento/magento2#18927]
### Preconditions (*)
1. Magento 2.3.0-alpha to beta 36
2. Payment Extension with an EntityPool item name, which has a minimum of 22 characters (generated sequence table name must be longer than 32 characters).

### Steps to reproduce (*)
1. Run setup:upgrade with 

### Expected result (*)
1. setup:upgrade process runs without any problems

### Actual result (*)
1. Unique constraint violation exception thrown in 'schema recurring' update step of Magento_SalesSequence module 

### Description
Having a too small length for the sequence_table column can lead to integrity constraint errors on setup:upgrade, because longer table names will be cut off (at least the store id suffix of those sales sequence table names).
In our case it was the PostfinanceCW extension, which lead to the tables "sequence_postfinancecw_transaction_0", "sequence_postfinancecw_transaction_1",... and so on (for each store id one table), which are 36 characters long.
So the names for the admin store view (id 0) got cut off in the sales_sequence_meta table and as a consequence lead to an integrity constrained, when the recurring script in the Magento_SalesSequence module tries to add the meta data for the sequence_postfinancecw_transaction_1 table.

**Comment by magento-engcom-team** [comment]
Hi @jhruehl. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@jhruehl do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by jhruehl** [comment]
So I already created a pull request during the 2.3.0-alpha release:
https://github.com/magento/magento2/pull/17387

This doesn't apply anymore, but the fix is more or less the same. The pull request was rejected, because @orlangur wanted the "getTableName" functionality to be used as a fix. Because neither him, especially not me (because the effort in this particular case was in no comparison to the benefits; and I don't work for free) and obviously no one else actually wanted to invest time to do so, nothing was done and the bug still exists.

So instead of making a fuss, please just increase the length of the sequence_table column to something reasonable like 128. The table at hand will never in no store setup ever reach a size where this actually would matter a tiny bit concerning the perfomance of the store. Especially since mysql versions, in which it would have mattered (a tiny bit), aren't usable anymore anyway.

**Comment by orlangur** [comment]
@jhruehl,

> because the effort in this particular case was in no comparison to the benefits

Nope, we should use intended mechanisms instead of hacky solutions here and there.

> and I don't work for free

This is irrelevant and does not justify poor implementation.

> In our case it was the PostfinanceCW extension, which lead to the tables "sequence_postfinancecw_transaction_0", "sequence_postfinancecw_transaction_1",... and so on (for each store id one table), which are 36 characters long.

Looks like the problem resides in third-party extension, isn't it? Anyway, putting `getTableName` in a proper place should not require too much efforts.

**Comment by jhruehl** [comment]
@orlangur I don't think this to be a hacky solution, because it actually doesn't change anything of the behaviour given. If you look closer it is actually just an increase of a parameter, which was obviously chosen to have a to low value.

Furthermore the extension causing the problem works perfectly fine in Magento 2.2.6 and I'm not sure how many other payment extensions will cause the problem once Magento 2.3.0 is released.

I don't say that you shouldn't find a better solution. I'm just saying that this is initially better than having no fix at all in place when Magento 2.3.0 is released. And given the time frame I don't think there will.

So because there are probably other much more relevant issues to fix at hand my proposed fix would at least not mess up several shops using external payment extensions when updating to Magento 2.3.0, which have extensions causing longer sales_sequence table names generated than currently possible.

I agree. Hacky solutions aren't wanted, because most cause other problems given other user cases. And if the solution would have been of the kind that, I would just have created an issue, post my current patch and wait several releases until my patch isn't needed anymore.

**Comment by heldchen** [comment]
I think no one disagrees with the notion that `getTableName` should be used. unfortunately, not all thrid party extensions are following best practizes.

our customers currently are using payment extensions that have long names, and as such run into this problem once upgraded to 2.3. for those customers that already have a production magento 2.2 store, chances are basically zero that the extension providers properly cleanup their mess in a 2.3 compatible version. altering table names can be messy - and even dangerous in database clusters depending on the table sizes -, especially when the devs already are proven to be not that competent (or else they would have used the default magento way for their tables in the first place).

as the current limit of 32 chars is rather arbitrary, allowing larger values while advocating for better third party extension code is lot less dangerous and might save a few of magento's customers some trouble.

**Comment by orlangur** [comment]
> chances are basically zero that the extension providers properly cleanup their mess in a 2.3 compatible version

No 2.2 extension should be used with 2.3 until it is thoroughly tested by vendor and all arisen bugs are fixed.

> as the current limit of 32 chars is rather arbitrary

This is not true, it was introduced long time ago during MMDB M1 project for compatibility with MSSQL/Oracle. Does Magento have any longer table names currently?

**Comment by heldchen** [comment]
> This is not true, it was introduced long time ago during MMDB M1 project for compatibility with MSSQL/Oracle. Does Magento have any longer table names currently?

sure, "32 characters" probably wasn't arbitrary at one time in the past. there are _lot_ of tables in a vanilla installation with longer names now. some highlights:
`catalog_product_entity_media_gallery_value_to_entity`
`catalog_product_entity_media_gallery_value_video`
`catalog_product_index_price_cfg_opt_agr_idx`

and in EE you seem to get some even longer tables:
`magento_customercustomattributes_sales_flat_order_address`
`magento_customercustomattributes_sales_flat_quote_address`

I'm not here to debate table name lengths tho, but rather to point out real world problems with the limit in `sales_sequence_meta`, a table which was introduced in magento 2. the extension provider is actually using magento's own `\Magento\SalesSequence\Model\Builder` class to create the sequences: 
```
		foreach (array_keys($this->_storeManager->getStores(true)) as $storeId) {
			$this->_sequenceBuilder->setPrefix($this->_sequenceConfig->get('prefix'))
				->setSuffix($this->_sequenceConfig->get('suffix'))
				->setStartValue($this->_sequenceConfig->get('startValue'))
				->setStoreId($storeId)
				->setStep($this->_sequenceConfig->get('step'))
				->setWarningValue($this->_sequenceConfig->get('warningValue'))
				->setMaxValue($this->_sequenceConfig->get('maxValue'))
				->setEntityType('postfinancecw_transaction')->create();
		}
```
In your previous post you argued the extension should shorten the name when in fact it is magento's `Magento_SalesSequence` module itself that creates the tables with names that are longer than its own imposed limits. moreover, the class _is_ already using your proposed solution `getTableName`:
```
    protected function getSequenceName()
    {
        return $this->appResource->getTableName(
            sprintf(
                'sequence_%s_%s',
                $this->data['entity_type'],
                $this->data['store_id']
            )
        );
    }
```
further down in the call stack `Magento\Framework\DB\Adapter\Pdo\Mysql::getTableName` is using `ExpressionConverter::shortenEntityName` which sets the limit to `MYSQL_IDENTIFIER_LEN`. and guess what: `    const MYSQL_IDENTIFIER_LEN = 64;`

so clearly, the length of the field `sequence_table` in the table `sales_sequence_meta` shouldn't be 32 but 64 chars.

**Comment by heldchen** [comment]
side note: the problem why this only started to show up now is related to the `db_schema.xml` introduction. 

the extension provider "fixed" the root cause by changing the field length to 64 chars his `InstallData` script (ugh!). while that might have worked well enough in the past, since 2.3 the table definition is "corrected" back to 32 chars through the DB Schema XML (one of my fav new tech!). mysql happily obliged by cutting the existing values in the `sequence_table` field.

**Comment by steffen-wirth** [comment]
i can confirm that this is still an issue with the latest release of v. 2.3.0 I tried to change the column type but with no luck. Is that because of the new db_schema.xml? is there an easy way to overwrite those db_schema.xml?

**Comment by orlangur** [comment]
@heldchen thanks, you're absolutely right. Too bad I didn't check existing limit thinking it was 32.

@steffen-wirth you can change this for your instance similar to PR: https://github.com/magento/magento2/pull/17387/files

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97361` were created

**Comment by ghost** [comment]
HI @jhruehl they already have a Pull request -> https://github.com/magento/magento2/pull/17387

**Comment by engcom-Charlie** [comment]
Hello @jhruehl 
Your issue is already fixed in `2.4-develop` branch.
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/SalesSequence/etc/db_schema.xml#L44

So i have to close this issue as not reproducible.

Please feel free to comment, reopen or create new ticket according to the Issue reporting guidelines
if you are still facing this issue on the latest Magento 2.4-develop branch.

Thank you for collaboration.
